tempesta_fw/ss_skb.c

```diff
@@ -386,21 +446,32 @@ __split_linear_data(struct sk_buff *skb, char *pspt, int len, TfwStr *it)
  * can be zero, but it can not be equal to the size of fragment @i.
  *
  * @return 0 on success, -errno on failure.
- * @return SKB in @it->skb if new SKB is allocated.
  * @return pointer to the room for new data in @it->ptr.
- * @return @it->flags is set if @it->ptr points to data in it->skb.
+ * @return pointer to SKB with data at @it->ptr in @it->skb.
  */
 static int
-__split_pgfrag_add(struct sk_buff *skb, int i, int off, int len, TfwStr *it)
+__split_pgfrag_add(struct sk_buff *skb, struct sk_buff *bskb,
+		   int i, int off, int len, TfwStr *it)
 {
 	int tail_len;
-	struct sk_buff *skb_dst;
+	struct sk_buff *skb_dst, *skb_new;
 	skb_frag_t *frag_dst, *frag = &skb_shinfo(skb)->frags[i];
 
 	SS_DBG("[%d]: %s: skb [%p] i [%d] off [%d] len [%d] fragsize [%d]\n",
 		smp_processor_id(), __func__,
 		skb, i, off, len, skb_frag_size(frag));
 	/*
+	 * Fast path: just advance the skb tail pointer.
+	 * If @off is zero and @i is zero (first paged fragment),
+	 * then see if there's sufficient room in the linear part.
+	 */
+	if (unlikely(!off && !i && (len <= ss_skb_tailroom(skb)))) {
+		ss_bskb_adjust_len(skb, bskb, len);
+		it->ptr = ss_skb_put(skb, len);
+		it->skb = bskb;
+		return 0;
+	}
```

Comment:
If I understand correctly you moved the quick path here from `__split_linear_data()` just because of `offset < d_size` in `__skb_fragment()`. IIRC we discussed the condition and agreed that we should use `<=`.